api/src/main/java/org/pmiops/workbench/dataset/DataSetServiceImpl.java

```diff
@@ -827,21 +815,185 @@ private String generateRandomEightCharacterQualifier() {
+      case HAIL:
+        return generateHailCode(qualifier, dataSetExportRequest);
+      case PLINK:
+        return generatePlinkCode(qualifier, dataSetExportRequest);
+      case NONE:
+        return generateDownloadVcfCode(qualifier, dataSetExportRequest);
+      default:
+        throw new BadRequestException("Invalid Genomics Analysis Tool");
+    }
+  }
+
+  private Optional<String> getExtractionDirectory(Long datasetId) {
+    try {
+      return Optional.of(
+              submissionDao
+                  .findMostRecentValidExtractionByDataset(dataSetDao.findById(datasetId).get())
+                  .get()
+                  .getOutputDir()
+                  .replaceFirst("/$", "")) // Drop trailing slash if exists
+          .filter(dir -> !dir.isEmpty());
+    } catch (NoSuchElementException e) {
+      return Optional.empty();
+    }
+  }
+
+  private String generateVcfDirEnvName(String qualifier) {
+    return "DATASET_" + qualifier + "_VCF_DIR";
+  }
+
+  private static final String MISSING_EXTRACTION_DIR_PLACEHOLDER =
+      "\"GCS_VCF_DIRECTORY_GOES_HERE\"";
+
+  private String generateExtractionDirCode(
+      String qualifier, DataSetExportRequest dataSetExportRequest) {
+    String extractionDir =
+        getExtractionDirectory(dataSetExportRequest.getDataSetRequest().getDataSetId())
+            .orElse(MISSING_EXTRACTION_DIR_PLACEHOLDER);
+
+    String noExtractionDirComment =
+        MISSING_EXTRACTION_DIR_PLACEHOLDER.equals(extractionDir)
+            ? "# VCF files for this dataset do not exist\n"
+                + "# Run a Genomic Extraction from a Dataset to generate VCF files\n"
+            : "";
+
+    return noExtractionDirComment
+        + "%env "
+        + generateVcfDirEnvName(qualifier)
+        + "="
+        + extractionDir;
+  }
+
+  private List<String> generateHailCode(
+      String qualifier, DataSetExportRequest dataSetExportRequest) {
+    final String matrixName = "mt_" + qualifier;
+
+    return ImmutableList.of(
+        "# Note: Hail demos must be run from a \"Hail Genomics Analysis\" cloud analysis environment",
+        generateExtractionDirCode(qualifier, dataSetExportRequest),
+        "# Initialize Hail\n"
+            + "\n"
+            + "import hail as hl\n"
+            + "import os\n"
+            + "from hail.plot import show\n"
+            + "\n"
+            + "hl.init(default_reference='GRCh38')\n"
+            + "hl.plot.output_notebook()",
+        "# Create Hail Matrix table\n"
+            + "# This can take a few hours for larger (several hundred participant) cohorts\n"
+            + "workspace_bucket = os.environ['WORKSPACE_BUCKET']\n"
+            + "vcf_dir = os.environ['"
+            + generateVcfDirEnvName(qualifier)
+            + "']\n"
+            + "hail_matrix_table_gcs = f'{workspace_bucket}/dataset_"
+            + qualifier
+            + ".mt'\n"
+            + "hl.import_vcf(f'{vcf_dir}/*.vcf.gz', force_bgz=True).write(hail_matrix_table_gcs)\n"
```

Comment:
Not for this PR, but maybe as a TODO here: it would be ideal if running this codecell can be more idempotent. e.g. check the "hail_matrix_table_gcs" directory to see whether it's been written already, skip the import if so.